src/base/math.jl

```diff
@@ -75,37 +77,41 @@ addtestN(:^, rand(1:10), rand(1:5))
 
 # clamp(x,lo,hi) clamps x between lo and hi
 @primitive clamp(x,lo,hi),dy,y  unbroadcast(x,dy.*(lo .<= x .<= hi))
-bf = broadcast_func(:clamp)
-if bf != :clamp
-    @eval @primitive $bf(x,d...),dy,y  unbroadcast(x,dy.*(d[1] .<= x .<= d[2]))
+let bf = broadcast_func(:clamp)
+    if bf != :clamp
+        @eval @primitive $bf(x,d...),dy,y  unbroadcast(x,dy.*(d[1] .<= x .<= d[2]))
+    end
+    addtest(:clamp, randn(), -1., 1.)
+    addtest(bf, randn(10), -1., 1.)
+    broadcast_func(:&)  # need this for (lo .<= x .<= hi)
 end
-addtest(:clamp, randn(), -1., 1.)
-addtest(bf, randn(10), -1., 1.)
-broadcast_func(:&)  # need this for (lo .<= x .<= hi)
 
 # ldexp(x,n) computes x*2^n with x real, n integer
 @primitive ldexp(x,n...),dy  (dy*(2.0^n[1]))
 addtest(:ldexp, randn(), rand(-2:2))
-bf = broadcast_func(:ldexp)
-if bf != :ldexp
-    @eval @primitive $bf(x,n...),dy  (dy.*(2.0^n[1]))    
-    addtest(bf, randn(2), rand(-2:2))
+let bf = broadcast_func(:ldexp)
+    if bf != :ldexp
+        @eval @primitive $bf(x,n...),dy  (dy.*(2.0^n[1]))    
+        addtest(bf, randn(2), rand(-2:2))
+    end
 end
 
 # mod2pi(x) returns modulus after division by 2pi for x real.
 @primitive mod2pi(x),dy dy
 addtest(:mod2pi, 100randn())
-bf = broadcast_func(:mod2pi)
-if bf != :mod2pi
-    @eval @primitive $bf(x),dy dy
-    addtest(bf, 100randn(2))
+let bf = broadcast_func(:mod2pi)
+    if bf != :mod2pi
+        @eval @primitive $bf(x),dy dy
+        addtest(bf, 100randn(2))
+    end
```

Comment:
Why do you use so many let expressions? Does this new structure have any advantage over the previous one? I mean this one is OK, you just want to hide something from the top level structure, but some of those let expressions happens in for loops. I guess you want to make all the parts convenient?

Reply:
If I remember correctly it was to get rid ofa  warning on 0.7 of "implicit assigning to global variable".